Eagerly create init(from:) when looking up 'init' on a Decodable type

jrose-apple opened the pull request:
Otherwise, the initializer won't be inherited properly onto a subclass, resulting in the base class being allocated instead of the subclass when using `Sub.init(from:)`.

[SR-8083](https://bugs.swift.org/browse/SR-8083) / rdar://problem/41426211

jrose-apple commented:
@swift-ci Please test

itaiferber commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -7950,12 +7950,11 @@ void TypeChecker::synthesizeMemberForLookup(NominalTypeDecl *target,
-    auto argumentName = argumentNames.front();
     if (baseName == DeclBaseName::createConstructor() &&
-        argumentName == Context.Id_from) {
+        (member.isSimpleName() || argumentNames.front() == Context.Id_from)) {
```
If `member` is a compound name, won't we fall into here regardless of what the argument name is?

itaiferber replied:
Unfortunately I haven't kept up with the `isSpecialName` changes, so correct me if I'm wrong, but if `member.isSimpleName()`, what makes it okay to ignore `argumentNames`? Wouldn't we end up falling into this case unnecessarily even when the argument name isn't `from:`? (That's what I'm trying to get at here.)

jrose-apple replied:
A simple name is like `encode` or `init`. A compound name is `encode(to:)` or `init(from:)`. So this code was doing the right synthesis whenever someone asked for `init(from:)` explicitly, but not if they just asked for `init`. The latter is what happens when trying to inherit initializers.

itaiferber replied:
Got it, thanks for the clarification! Out of curiosity, then, is this conformance check not going to be expensive to perform on _every_ inheritance of an `init`? Why does inheritance not specify the full compound name?

jrose-apple replied:
This is trying to figure out what initializers there are at all, so the only thing it can look up is `init`. Indeed, because we can synthesize `init(from:)`, any subclassing action has to do this check.

itaiferber replied:
Okay, I guess there isn't much we can do about that here without changing the inheritance mechanism.

My last lingering question is where this inheritance check might happen relative to regular protocol conformance checking; is it possible for this to trigger a conformance check on a subclass before the superclass is ever checked? Would that correctly trigger a check on the superclass first?

itaiferber replied:
Excellent — thanks for fielding my concerns! Looks good once we figure out the source of these failures.

jrose-apple commented:
Uh, hm. That's not good.

jrose-apple committed:
Eagerly create init(from:) when looking up 'init' on a Decodable type

Otherwise, the initializer won't be inherited properly onto a
subclass, resulting in the base class being allocated instead of the
subclass when using Sub.init(from:).

https://bugs.swift.org/browse/SR-8083

jrose-apple commented:
Ah, just forgot about tests that don't import the stdlib.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please smoke test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
[Compiler performance smoke test here](https://ci.swift.org/view/Pull%20Request/job/swift-PR-compiler-performance-smoke-test-macOS/277/)
